test/data/html/tests.js

```diff
@@ -3839,6 +3839,304 @@ exports.test_data = {
         '</span>'
       ]
     }]
+  }, {
```

Comment:
These are a solid set of tests.  We need a couple more: 

* do not indent an unrecognized `@-string` - `@foo (true) {` should be treated as regular content.
* do not indent invalid syntax - `@if [] {` is missing the required open parenthesis, treat as regular content.
* do not match escaped `@` symbol - I don't know how angular escapes, but ... 
* do not match if `@-string` is not preceded by `}` or whitespace (including newline)

Reply:
> * do not match if `@-string` is not preceded by `}` or whitespace (including newline)

For this, could you please provide an example of what you mean?

Reply:
I believe this is valid angular control flow and should match and be indented:
```
FYI: I want you to know that @if (a > b) {
{{a}} is greater than {{b}}.
}@else{ 
{{a}} is less than or equal to {{b}}.
}

Or to say it another way, {{b}} @if (a > b) {is less than}@else{is greater than or equal to} {{a}}.
```

There's a bunch of missing spaces, but the @-strings are always preceded by a space (tab or newline would also be okay), or a `}`. 

The following might be valid control flow but are too close to other strings you might see in non-angular scenarios, so do not match or indent:
```
My email is bitwiseman@if.com (only for work). 
bitwiseman@if (true) { }
```

Reply:
> ```
> My email is bitwiseman@if.com (only for work). 
> bitwiseman@if (true) { }
> ```
For this, the code is not selecting any control flow token, as the `TK_TEXT` token selector is not being cut when a `@` character is found, only at spaces.
![image](https://github.com/beautify-web/js-beautify/assets/84864554/9392fd6c-810b-49b4-a47e-d29d96f15d7f)

> ```
> Or to say it another way, {{b}} @if (a > b) {is less than}@else{is greater than or equal to} {{a}}.
> ```
For this, I've found a bug in my code: the `@if (a > b) {` is registered as a control flow opening token, but the `}` character is not as a closing one (because it is a part of the `than}@else{is` TK_TEXT token.
When we implement the improvement of the templating option, we can improve the selectors and register the tokens in a better way, but for now I'm planning to just check for a `}` character : if currently a control flow is opened, it's a control flow closing character. Any other way: it's just text. I'm doing it today-tomorrow also.

Reply:
@bitwiseman I tried a lot of different approaches, but because the special characters (`{`, `}`, `@`) can have a lot of meanings in different templating languages (`@` may be just a character, while it's a reserved character in Angular; `{` and `}` may be block opening characters, handlebar characters or just simple characters), it's simply just not possible to make a 100% percent match for those characters, especially if they are not preceded by a whitespace character in the current implementation. 
Example:
We'd like to match this as an angular control flow syntax: `{{b}} @if (a > b) {is less than}@else{is greater than or equal to} {{a}}`.

But we'd like to not match the following as an angular control flow syntax: `bitwiseman@if (true) { }`.

Unfortunately, by code I cannot differentiate the two cases if it is unknown what templating language we are using. A really good match could be achieved if we could aim only for the specific templating language, hope we can do it one day, as you suggested.